### DISCUSSION TITLE
Are the C++ compilers able to optimize `x = conj(x)` when x is real?

### DISCUSSION THREAD

**Discussion by weslleyspereira** [tlapack/tlapack#43] [Q&A] [answered]
When `x` is not complex, `conj(x)` returns `x`. Sometimes we need to write `x = conj(x)`, e.g., `her2k, herk`. For instance:

https://github.com/tlapack/tlapack/blob/774ea0d9b9e770371edbbec9e7c57d73ce1380bb/include/blas/her2k.hpp#L216-L221

In this example, I expect the compiler avoid all those lines since `x = x` is trivial.

My question is: is it ok to assume this optimization will be carried by the compiler or should we write an in-place conjugate operator?

**Comment by thijssteel** [comment, marked as answer]
I'd definitely say yes. When writing expression templates, the compiler is generally able to sift through all that complicated code and optimize it to a high degree. If it can do that, it can definitely optimize away a trivial function.

**Comment by weslleyspereira** [comment]
I come back to this question. There is a simple code that shows that the answer could be actually No.

I tested the code below on https://godbolt.org/z/5vGdnT8qq. The GCC 12.2 compiler with flag -O3 gives me a different answer for:
1. keeping `C(i,j) = conj( C(j,i) );` (190 instructions), or
2. replacing this line by the trivial `C(i,j);` (89 instructions)

```c++
#include <iostream>
#include <complex>

template <typename T>
struct is_complex:
    std::integral_constant<bool, false>
{};
/// specialize for std::complex
template <typename T>
struct is_complex< std::complex<T> >:
    std::integral_constant<bool, true>
{};

template<class T, std::enable_if_t<!is_complex<T>::value,int> = 0 >
inline constexpr
T conj( const T& x ) { return x; }

#define C(i,j) C_[i+j*n]

int main() {
    using idx_t = size_t;
    const idx_t n = 100;
    float C_[n*n];

    for(idx_t j = 0; j < n; ++j) { 
        for(idx_t i = j+1; i < n; ++i) 
            C(i,j) = conj( C(j,i) ); 
    }

    float sum = 0;
    for(idx_t j = 0; j < n; ++j) { 
        for(idx_t i = j+1; i < n; ++i) 
            sum += std::abs(C(i,j));
    }

    return int( sum );
}
```

**Comment by weslleyspereira** [comment]
My bad here. I come back to this topic. I noticed I had changed j by i, and arrived to the wrong conclusion.

Since I couldn't reproduce what I saw in https://github.com/tlapack/tlapack/discussions/43#discussioncomment-3882394, I think we should usually expect the compiler to be able to optimize it.
